### PR TITLE
perf(core): Make `PlatformLocation` tree-shakable

### PR DESCRIPTION
**Pull request by mhevery**
Convert `PlatformLocation` into a tree-shakable provider

### CODE REVIEW

File: packages/compiler-cli/test/compliance/r3_compiler_compliance_spec.ts [hunk 1]
```diff
@@ -2785,7 +2785,7 @@ describe('compiler compliance', () => {
               $r3$.ɵɵelementStart(1, "div");
               $r3$.ɵɵtext(2);
               $r3$.ɵɵelementEnd();
-              $r3$.ɵɵelementStart(3, "ul");
+              $r3$.��ɵelementStart(3, "ul");
```
Comment (alfaproject) [inline packages/compiler-cli/test/compliance/r3_compiler_compliance_spec.ts]: 👀

Comment (mhevery) [reply packages/compiler-cli/test/compliance/r3_compiler_compliance_spec.ts]: Not ready for review yet. :-)

File: packages/core/src/di/scope.ts [hunk 2]
```diff
@@ -14,5 +14,4 @@ import {InjectionToken} from './injection_token';
  * as a root scoped injector when processing requests for unknown tokens which may indicate
  * they are provided in the root scope.
  */
-export const APP_ROOT = new InjectionToken<boolean>(
-    'The presence of this token marks an injector as being the root injector.');
+export const INJECTOR_SCOPE = new InjectionToken<'root'|'platform'|null>('Set Injector scope.');
```
Comment (alxhub) [inline packages/core/src/di/scope.ts]: Just FYI, I believe there may be some projects in g3 which make use of this private API.

File: packages/core/src/di/r3_injector.ts [hunk 3]
```diff
@@ -128,7 +130,8 @@ export class R3Injector {
 
     // Detect whether this injector has the APP_ROOT_SCOPE token and thus should provide
     // any injectable scoped to APP_ROOT_SCOPE.
-    this.isRootInjector = this.records.has(APP_ROOT);
+    const record = this.records.get(INJECTOR_SCOPE);
+    this.scope = record ? record.value : null;
```
Comment (alxhub) [inline packages/core/src/di/r3_injector.ts]: Use `record != null`, per @pkozlowski-opensource's perf talk about the cost of implicit conversions.

File: packages/core/src/di/r3_injector.ts [hunk 4]
```diff
@@ -83,8 +83,10 @@ export function createInjector(
 export class R3Injector {
   /**
    * Map of tokens to records which contain the instances of those tokens.
+   * - `null` value implies that we don't have the record. Used by tree-shakable injectors
```
Comment (alxhub) [inline packages/core/src/di/r3_injector.ts]: Clever! 👍

File: packages/platform-server/src/server.ts [hunk 5]
```diff
@@ -27,6 +27,8 @@ function notSupported(feature: string): Error {
   throw new Error(`platform-server does not support '${feature}'.`);
 }
 
+type __retain_for_correct_d_ts_generation__ = [PlatformRef];
```
Comment (alxhub) [inline packages/platform-server/src/server.ts]: Maybe a comment here describing what happens if this isn't here?

**Review by alxhub — APPROVED** [review]
Overall this LGTM.

The only thing I would suggest is this PR is one commit without much of a description, and it does a lot more than it says in the commit title. I would recommend breaking the PR into 2-3 commits (esp. one for the change to tree-shakeable provider infrastructure which allows platform providers). At the very least the commit description should document what changes were made to the overall infrastructure to allow the conversion of `PlatformLocation` to be tree-shakeable.

File: packages/core/src/di/scope.ts (hunk 2 above)
Comment (mhevery) [reply packages/core/src/di/scope.ts]: thanks for the heads up

File: packages/core/src/di/r3_injector.ts (hunk 3 above)
Comment (mhevery) [reply packages/core/src/di/r3_injector.ts]: fixed

File: packages/platform-server/src/server.ts (hunk 5 above)
Comment (mhevery) [reply packages/platform-server/src/server.ts]: I realized that the issue is that we rely on type inference. So I just properly typed the code and there is no need for this any more.

**Comment by mhevery** [comment]
@alxhub Thanks for the feedback. I have broken the PR into 3 SHAs.

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:266015776:BASE:266015817:1567033668112:7cc8a7fc)
[presubmit ivy](http://test/OCL:266015776:BASE:266016098:1567033759850:ddf6c5a)

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:266015776:BASE:266026091:1567036864109:deeb8b0b)
[presubmit ivy](http://test/OCL:266015776:BASE:266026761:1567037014301:d6e6c758)

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:266015776:BASE:266058971:1567052511675:4756a895)
[presubmit ivy](http://test/OCL:266015776:BASE:266058971:1567052511675:4756a895)

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:266226030:BASE:266234225:1567115271547:8e3a6375)
[presubmit clean](http://test/OCL:266226030:BASE:266232844:1567114940361:e9030330)

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:266226030:BASE:266244387:1567118035327:e2aed72)
[presubmit clean](http://test/OCL:266226030:BASE:266246602:1567118678479:9f81aba9)

**Comment by mhevery** [comment]
MERGE ASSISTANCE: needs cl/266062693 patch

**Comment by razvanm** [comment]
As far as I can tell this PR broke the overriding the `Location.stripTrailingSlash` mentioned in https://github.com/angular/angular/issues/16051#issuecomment-330465196. Is that expected?

**Comment by mhevery** [comment]
@razvanm What specifically broke. We just moved code around, there should be no changes in behavior.

**Comment by razvanm** [comment]
In the code I'm working with I had in my root app the following code:

```
Location.stripTrailingSlash = (url: string) => {
  return url.endsWith('/') && url.length > 1 ? url + '.' : url;
};
```

This used to work fine both in dev mode and in the production code that gets compiled using the Closure compiler. After this PR the overriding stopped working (I used some `console.log()` calls to figure out that the Angular code is called instead of mine).

**Comment by mhevery** [comment]
Right, you are using unsupported API. We don't guarantee that you can
override our API, it is not part of our public contract.

**Pull request closed without merging**